In [5]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# Load Data

In [6]:
RAW_DATA_PATH = "../data/raw/survey_results_public.csv"
RAW_DATA_PATH_2020 = "../data/raw/2020/survey_results_public.csv"

In [7]:
raw_df = pd.read_csv(RAW_DATA_PATH_2020)
raw_df

Respondent                                         MainBranch Hobbyist  \
0               1                     I am a developer by profession      Yes   
1               2                     I am a developer by profession       No   
2               3                        I code primarily as a hobby      Yes   
3               4                     I am a developer by profession      Yes   
4               5  I used to be a developer by profession, but no...      Yes   
...           ...                                                ...      ...   
64456       64858                                                NaN      Yes   
64457       64867                                                NaN      Yes   
64458       64898                                                NaN      Yes   
64459       64925                                                NaN      Yes   
64460       65112                                                NaN      Yes   

        Age Age1stCode CompFreq  CompTotal  ConvertedComp             Country  \
0       NaN         13  Monthly        NaN            NaN             Germany   
1       NaN         19      NaN        NaN            NaN      United Kingdom   
2       NaN         15      NaN        NaN            NaN  Russian Federation   
3      25.0         18      NaN        NaN            NaN             Albania   
4      31.0         16      NaN        NaN            NaN       United States   
...     ...        ...      ...        ...            ...                 ...   
64456   NaN         16      NaN        NaN            NaN       United States   
64457   NaN        NaN      NaN        NaN            NaN             Morocco   
64458   NaN        NaN      NaN        NaN            NaN            Viet Nam   
64459   NaN        NaN      NaN        NaN            NaN              Poland   
64460   NaN        NaN      NaN        NaN            NaN               Spain   

         CurrencyDesc  ...                  SurveyEase           SurveyLength  \
0       European Euro  ...  Neither easy nor difficult  Appropriate in length   
1      Pound sterling  ...                         NaN                    NaN   
2                 NaN  ...  Neither easy nor difficult  Appropriate in length   
3        Albanian lek  ...                         NaN                    NaN   
4                 NaN  ...                        Easy              Too short   
...               ...  ...                         ...                    ...   
64456             NaN  ...                         NaN                    NaN   
64457             NaN  ...                         NaN                    NaN   
64458             NaN  ...                         NaN                    NaN   
64459             NaN  ...                         NaN                    NaN   
64460             NaN  ...                         NaN                    NaN   

      Trans                                     UndergradMajor  \
0        No  Computer science, computer engineering, or sof...   
1       NaN  Computer science, computer engineering, or sof...   
2       NaN                                                NaN   
3        No  Computer science, computer engineering, or sof...   
4        No  Computer science, computer engineering, or sof...   
...     ...                                                ...   
64456   NaN  Computer science, computer engineering, or sof...   
64457   NaN                                                NaN   
64458   NaN                                                NaN   
64459   NaN                                                NaN   
64460   NaN  Computer science, computer engineering, or sof...   

            WebframeDesireNextYear                      WebframeWorkedWith  \
0                     ASP.NET Core                    ASP.NET;ASP.NET Core   
1                              NaN                                     NaN   
2                              NaN                           

In [8]:
# Get the data from a random Respondent

raw_df.sample(1).T

4623
Respondent                                             4639
MainBranch           I am a student who is learning to code
Hobbyist                                                Yes
Age                                                    18.0
Age1stCode                                               12
...                                                     ...
WebframeWorkedWith                                      NaN
WelcomeChange       Just as welcome now as I felt last year
WorkWeekHrs                                             NaN
YearsCode                                                 6
YearsCodePro                                            NaN

[61 rows x 1 columns]

In [9]:
raw_df.shape

(64461, 61)

In [15]:
raw_df.info() 


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64461 entries, 0 to 64460
Data columns (total 61 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Respondent                    64461 non-null  int64  
 1   MainBranch                    64162 non-null  object 
 2   Hobbyist                      64416 non-null  object 
 3   Age                           45446 non-null  float64
 4   Age1stCode                    57900 non-null  object 
 5   CompFreq                      40069 non-null  object 
 6   CompTotal                     34826 non-null  float64
 7   ConvertedComp                 34756 non-null  float64
 8   Country                       64072 non-null  object 
 9   CurrencyDesc                  45472 non-null  object 
 10  CurrencySymbol                45472 non-null  object 
 11  DatabaseDesireNextYear        44070 non-null  object 
 12  DatabaseWorkedWith            49537 non-null  object 
 13  D

In [16]:
# This is an alternative of the previous cell. 
# The aim: calculate the number of NULL values
raw_df.isnull().sum().sort_values(ascending=False)

ConvertedComp             29705
CompTotal                 29635
NEWPurchaseResearch       27140
PurchaseWhat              25097
WebframeDesireNextYear    24437
                          ...  
Employment                  607
Country                     389
MainBranch                  299
Hobbyist                     45
Respondent                    0
Length: 61, dtype: int64

In [18]:
# Calculate the portion of null values
raw_df.isnull().sum().sort_values(ascending=False) / raw_df.shape[0]

ConvertedComp             0.460821
CompTotal                 0.459735
NEWPurchaseResearch       0.421030
PurchaseWhat              0.389336
WebframeDesireNextYear    0.379097
                            ...   
Employment                0.009417
Country                   0.006035
MainBranch                0.004638
Hobbyist                  0.000698
Respondent                0.000000
Length: 61, dtype: float64

In [13]:
# Check the datatypes 
raw_df.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64461 entries, 0 to 64460
Data columns (total 61 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Respondent                    64461 non-null  int64  
 1   MainBranch                    64162 non-null  object 
 2   Hobbyist                      64416 non-null  object 
 3   Age                           45446 non-null  float64
 4   Age1stCode                    57900 non-null  object 
 5   CompFreq                      40069 non-null  object 
 6   CompTotal                     34826 non-null  float64
 7   ConvertedComp                 34756 non-null  float64
 8   Country                       64072 non-null  object 
 9   CurrencyDesc                  45472 non-null  object 
 10  CurrencySymbol                45472 non-null  object 
 11  DatabaseDesireNextYear        44070 non-null  object 
 12  DatabaseWorkedWith            49537 non-null  object 
 13  D

In [19]:
NUM_FEATS = ["Age", "YearsCode", "YearsCodePro", "OrgSize"]
raw_df[NUM_FEATS].sample().iloc[0]

Age             21.0
YearsCode          2
YearsCodePro     NaN
OrgSize          NaN
Name: 50247, dtype: object

In [20]:
INV_FEATS = ['YearsCode', 'YearsCodePro'] # Features to check (why they datatype is object. They are supposed to be numerical)
for inv_feat in INV_FEATS:
    unique_values = raw_df[inv_feat].unique()
    print(f".. {inv_feat}: {unique_values}")
    print(f"--------------------------------")


.. YearsCode: ['36' '7' '4' '15' '6' '17' '8' '10' '35' '5' '37' '19' '9' '22' '30' '23'
 '20' '2' 'Less than 1 year' '3' '13' '25' '16' '43' '11' '38' '33' nan
 '24' '21' '12' '40' '27' '50' '46' '14' '18' '28' '32' '44' '26' '42'
 '31' '34' '29' '1' '39' '41' '45' 'More than 50 years' '47' '49' '48']
--------------------------------
.. YearsCodePro: ['27' '4' nan '8' '13' '2' '7' '20' '1' '23' '3' '12' '17' '18' '10' '14'
 '29' '6' '28' '9' '15' '11' '16' '25' 'Less than 1 year' '5' '21' '19'
 '35' '24' '32' '22' '30' '38' '26' '40' '33' '31' 'More than 50 years'
 '34' '36' '39' '37' '41' '45' '47' '42' '46' '50' '43' '44' '48' '49']
--------------------------------


In [33]:
for feat in NUM_FEATS:
    unique_values = raw_df[feat].unique()
    print(f".. {feat}: {unique_values}")
    print(f"--------------------------------")

.. Age: ['Under 18 years old' '35-44 years old' '45-54 years old'
 '18-24 years old' '25-34 years old' '55-64 years old' 'Prefer not to say'
 '65 years or older']
--------------------------------
.. YearsCode: [nan '20' '37' '4' '9' '10' '7' '1' '15' '30' '31' '6' '12' '22' '5' '36'
 '25' '44' '24' '18' '3' '8' 'More than 50 years' '11' '29' '40' '39' '2'
 '42' '34' '19' '35' '16' '33' '13' '23' '14' '28' '17' '21' '43' '46'
 '26' '32' '41' '45' '27' '38' '50' '48' '47' 'Less than 1 year' '49']
--------------------------------
.. YearsCodePro: [nan '17' '27' '7' '11' '25' '12' '10' '3' 'Less than 1 year' '18' '37'
 '15' '20' '6' '2' '16' '8' '14' '4' '45' '1' '24' '29' '5' '30' '26' '9'
 '33' '13' '35' '23' '22' '31' '19' '21' '28' '34' '32' '40' '50' '39'
 '44' '42' '41' '36' '38' 'More than 50 years' '43' '47' '48' '46' '49']
--------------------------------
.. OrgSize: [nan '100 to 499 employees' '2 to 9 employees'
 'Just me - I am a freelancer, sole proprietor, etc.' '10 to 19 empl

In [23]:
for feat in raw_df.columns.values: 
    if "Worked" in feat:
        print(feat)
        
        

DatabaseWorkedWith
LanguageWorkedWith
MiscTechWorkedWith
NEWCollabToolsWorkedWith
PlatformWorkedWith
WebframeWorkedWith


In [25]:
FEATS = ["DatabaseWorkedWith", 
         "LanguageWorkedWith", 
         "MiscTechWorkedWith", 
         "NEWCollabToolsWorkedWith", 
         "PlatformWorkedWith", 
         "WebframeWorkedWith"
         ]

LABEL = ["DevType"]

raw_df[FEATS + LABEL].sample(1).iloc[0]

DatabaseWorkedWith                            DynamoDB;MongoDB;MySQL;Redis
LanguageWorkedWith          Bash/Shell/PowerShell;Go;Java;Python;Scala;SQL
MiscTechWorkedWith                    Apache Spark;Node.js;Pandas;Teraform
NEWCollabToolsWorkedWith                           Confluence;Github;Slack
PlatformWorkedWith            AWS;Docker;Google Cloud Platform;Linux;MacOS
WebframeWorkedWith                                                     NaN
DevType                          Engineer, data;Engineer, site reliability
Name: 3388, dtype: object

In [26]:
raw_df['DevType'].unique()

array(['Developer, desktop or enterprise applications;Developer, full-stack',
       'Developer, full-stack;Developer, mobile', nan, ...,
       'Data scientist or machine learning specialist;Database administrator;Developer, back-end;Developer, front-end;System administrator',
       'Academic researcher;Data or business analyst;Data scientist or machine learning specialist;Database administrator;Designer;Developer, back-end;Developer, desktop or enterprise applications;Developer, embedded applications or devices;Developer, front-end;Developer, full-stack;Developer, game or graphics;Developer, mobile;Developer, QA or test;DevOps specialist;Senior executive/VP;System administrator',
       'Data or business analyst;Developer, mobile;Senior executive/VP;System administrator'],
      dtype=object)

In [27]:
# The portion of students in the dataset 
# Student is not a job. We will delete this DevType 

students_df = raw_df[raw_df['DevType'] == 'Student']

print("Portion of students in the dataset: ")
print(f".. {students_df.shape[0] / raw_df.shape[0] * 100:.2f}%")

Portion of students in the dataset: 
.. 0.00%


In [28]:
FEATS = ["DatabaseWorkedWith", 
         "LanguageWorkedWith", 
         "MiscTechWorkedWith", 
         "NEWCollabToolsWorkedWith", 
         "PlatformWorkedWith", 
         "WebframeWorkedWith"
         ]

LABEL = ["DevType"]

OTHER_FEATS = []
for feature in raw_df.columns:
    if feature not in FEATS + LABEL:
        OTHER_FEATS.append(feature)
        
raw_df[OTHER_FEATS].sample(1).iloc[0]


Respondent                                                                  10992
MainBranch                                         I am a developer by profession
Hobbyist                                                                      Yes
Age                                                                          24.0
Age1stCode                                                                     10
CompFreq                                                                  Monthly
CompTotal                                                                     NaN
ConvertedComp                                                                 NaN
Country                                                                    Turkey
CurrencyDesc                                                         Turkish lira
CurrencySymbol                                                                TRY
DatabaseDesireNextYear                                Firebase;MongoDB;PostgreSQL
EdLevel         

In [29]:
unique_main_branch = raw_df['MainBranch'].unique()
print(f"unique_main_branch: {unique_main_branch}")

unique_main_branch: ['I am a developer by profession' 'I code primarily as a hobby'
 'I used to be a developer by profession, but no longer am'
 'I am not primarily a developer, but I write code sometimes as part of my work'
 'I am a student who is learning to code' nan]


In [30]:
PRO_MAIN_BRANCH = ["I am a developer by profession", 
                   "I am not primarily a developer, but I write code sometimes as part of my work/studies", 
                   "I used to be a developer by profession, but no longer am"
                   ]

portion_sum = 0
for branch in PRO_MAIN_BRANCH:
    print(f"Portion of {branch} in the dataset")
    print(f".. {branch}: {raw_df[raw_df['MainBranch'] == branch].shape[0] / raw_df.shape[0] * 100:.2f}%")
    portion_sum += raw_df[raw_df['MainBranch'] == branch].shape[0] / raw_df.shape[0] * 100
    
print(f"------------------------------------------")
print(f"The sum of these portions is: {portion_sum:.2f}%")
print(f"The portion of people who code as a hobby: {(100 - portion_sum):.2f}%")

Portion of I am a developer by profession in the dataset
.. I am a developer by profession: 73.21%
Portion of I am not primarily a developer, but I write code sometimes as part of my work/studies in the dataset
.. I am not primarily a developer, but I write code sometimes as part of my work/studies: 0.00%
Portion of I used to be a developer by profession, but no longer am in the dataset
.. I used to be a developer by profession, but no longer am: 1.84%
------------------------------------------
The sum of these portions is: 75.05%
The portion of people who code as a hobby: 24.95%



These values insure as that the dataset could be used in our business case. 
- We have only 7% of students in the dataset
- Among professionals, we have 76% of developers that write code as part of their job and only 11% of programmers that code as a hobby
These values indicates that the tech stacks will be correlated with the DevType. In addition, they reflect a real relationship.

END OF NOTEBOOK